### ISSUE TITLE
Standardize the CL interface

### ISSUE THREAD

**Issue by nsoblath** [project8/scarab#37]
The CL parser should use the standard POSIX interface instead of the rather customized interface it uses now.  I'll follow this up with some information on what CL parsing libraries are available and might work.

**Comment by laroque** [comment]
[An option to look at](https://github.com/hbristow/argparse), seems inspired by the python library I use, is header-only. Haven't looked carefully.

**Comment by nsoblath** [comment]
## How it would be used

```
command --knownopt=value1 nonoptarg=value2 nonoptarg2=value3 . . .
```

The executable (or a library, if we can reasonably duplicate Nymph's initialization-stage setup) will define known options that will be applied in the executable (or the relevant part of a library). 

Non-option arguments will be treated as they are in the current system and merged into the config.

## Requirements

* Adheres to GNU CL option standards
* Allows non-option arguments

## Would be nice

* Header only
* Fully cross-platform

## Evaluation of available libraries

* Program Options Parser Library (popl) -- http://github.com/badaix/popl
  * Header only & no external dependencies
  * Cross-platform
  * Adheres to GNU standard
  * Automatically creates usage message
  * Allows unknown options
  * Known options exist as objects and can be assigned directly to variables
  * Unknown and non-option arguments are available from the parser object
  * Comment: looks pretty good
  * 👍 

* The Lean Mean C++ Option Parser -- http://optionparser.sourceforge.net
  * Header only & no external dependencies
  * Cross-platform
  * Adheres to GNU standard
  * Generates usage message
  * Allows unknown options
  * Options are defined in some weird array or struct format
  * Comment: satisfies requirements, but is not aesthetically pleasing in some way I'm having trouble defining
  * ❔ 

* Boost Program Options -- http://www.boost.org --> program_options
  * No external dependencies (outside of some basic boost libraries)
  * Cross-platform
  * Adheres to GNU standard
  * Generates usage message
  * Allows unknown options
  * Comment: seems overly complicated, but would probably work
  * ❔ 

* Command line parser -- http://www.codeproject.com/Articles/2773/Command-line-parser
  * Windows oriented
  * 👎 

* Templatized C++ Command Line Argument Parser (TCLAP) -- http://tclap.sourceforge.net
  * Header only & no external dependencies
  * Cross-platform
  * Unclear if it supports non-option arguments
  * 👎 

* cxxopts -- http://github.com/jarro2783/cxxopts
  * Header only & no external dependencies
  * Allows non-option arguments, called "positional"
  * Add options directly to the parser object
  * Results accessible in map form
  * Does not allow unknown options
  * Comment: looks pretty nice, though I don't yet understand the non-option argument syntax
  * 👍 

* arg -- http://ccdw.org/~cjj/prog/arg
  * Header only & no external dependencies
  * Allows non-option arguments
  * Add options to parser
  * Access from variables where the option is "stowed"
  * Generates usage message
  * 👍 

* Argtable3 -- http://www.argtable.org
  * No external dependencies
  * Cross platform
  * Uses a seemingly-complicated setup of structs
  * 👎 

* getopt -- comes with GNU c installations
  * Header only & no external dependencies
  * Allows non-option arguments
  * Would require more work than many of the other options
  * 👎 

* argparse -- http://github.com/hbristow/argparse
  * Header only & no external dependencies
  * Allows non-option "final" arguments
  * Generates usage message but no help strings
  * Add options to parser
  * Access with map-like interface
  * Comment: very straightforward interface, but lacks help strings
  * 👍

**Comment by nsoblath** [comment]
So the acceptable options in my completely qualitative ranking: 

1. popl
1. cxxopts
1. argparse
1. arg

So the initial attempt at this work will use the popl library.

**Comment by nsoblath** [comment]
## Some implementation thoughts mostly for my own future benefit

I'd like to build in the ability to add options in a library that is linked to the executable using the options and the scarab library.  This means we'll want to have a global parser object, presumably the parser object provided by the parsing library wrapped in a singleton class.

There will need to be a class or classes of which instances can be created at static initialization time to create all of the various argument objects.  The constructor for those classes will access the singleton parser-wrapping object.

We'll also need to have access to those objects to get values out.  I see three ways this is doable:
1. Since these are global static objects within their compilation units, they will be visible there.  So that might be sufficient.  We should then make sure the variable names are reasonable.  
2. Another option would be to have those objects available via the singleton class that wraps the parser.  
3. Or we could go with the cxxopts library that already makes option values available via the parser.

**Comment by nsoblath** [comment]
The majority of this has now been implemented, as of 6d41f47d9c148faa459f88bbd482d1d7581bbaae.  The CLI11 library was used.

The ability to add options from other parts of the code has not been added yet, and will be moved to a new issue.
